cmd/compute-domain-daemon/dnsnames.go

```diff
@@ -0,0 +1,221 @@
+
+	// Make a local copy of the current ipToDNSName mappings
+	ipToDNSName := maps.Clone(m.ipToDNSName)
+
+	// Prefilter nodes to only consider those with the matching cliqueID
+	var cliqueNodes []*nvapi.ComputeDomainNode
+	for _, node := range nodes {
+		if node.CliqueID == m.cliqueID {
+			cliqueNodes = append(cliqueNodes, node)
+		}
+	}
+
+	// Find and remove stale IPs from map
+	currentIPs := make(map[string]bool)
+	for _, node := range cliqueNodes {
+		currentIPs[node.IPAddress] = true
+	}
+	for ip := range ipToDNSName {
+		if !currentIPs[ip] {
+			delete(ipToDNSName, ip)
+		}
+	}
+
+	// Add new IPs to map
+	for _, node := range cliqueNodes {
+		// If IP already has a DNS name, skip it
+		if _, exists := ipToDNSName[node.IPAddress]; exists {
+			continue
+		}
+
+		// Construct the DNS name from the node index
+		dnsName, err := m.constructDNSName(node)
+		if err != nil {
+			return fmt.Errorf("failed to allocate DNS name for IP %s: %w", node.IPAddress, err)
+		}
+
+		// Assign the IP -> DNS name mapping
+		ipToDNSName[node.IPAddress] = dnsName
+	}
+
+	// If the existing ipToDNSName mappings are unchanged, exit early
+	if maps.Equal(ipToDNSName, m.ipToDNSName) {
+		return nil
+	}
+
+	// Otherwise, update the cached ipToDNSName mapping
+	m.ipToDNSName = ipToDNSName
+
+	// And updated the hosts file with new mappings
+	return m.updateHostsFile()
+}
+
+// LogDNSNameMappings logs the current compute-domain-daemon mappings from memory.
+func (m *DNSNameManager) LogDNSNameMappings() {
+	m.Lock()
+	defer m.Unlock()
+
+	if len(m.ipToDNSName) == 0 {
+		klog.Infof("No compute-domain-daemon mappings found")
+		return
+	}
+
+	klog.Infof("Current compute-domain-daemon mappings:")
+	for ip, dnsName := range m.ipToDNSName {
+		klog.Infof("  %s -> %s", ip, dnsName)
+	}
+}
+
+// contructDNSName constructs a DNS name for a node based on its index field.
+// Returns an error if the index is invalid or exceeds maxNodesPerIMEXDomain.
+func (m *DNSNameManager) constructDNSName(node *nvapi.ComputeDomainNode) (string, error) {
+	if node.Index < 0 {
+		return "", fmt.Errorf("node %s has invalid index %d", node.Name, node.Index)
+	}
+	if node.Index >= m.maxNodesPerIMEXDomain {
+		return "", fmt.Errorf("node %s has invalid index %d, must be less than %d", node.Name, node.Index, m.maxNodesPerIMEXDomain)
+	}
+	dnsName := fmt.Sprintf(dnsNameFormat, node.Index)
+	return dnsName, nil
+}
+
+// updateHostsFile updates the /etc/hosts file with current IP to DNS name mappings.
+func (m *DNSNameManager) updateHostsFile() error {
+	// Read hosts file
+	hostsContent, err := os.ReadFile(hostsFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to read %s: %w", hostsFilePath, err)
+	}
+
+	// Grab any lines to preserve, skipping existing DNS name mappings
+	var preservedLines []string
+	for _, line := range strings.Split(string(hostsContent), "\n") {
+		line = strings.TrimSpace(line)
+
+		// Keep empty lines and comments
+		if line == "" || strings.HasPrefix(line, "#") {
```

Comment:
Can this entire condition go away?
Should we not simply retain all lines that do not include our naming pattern?